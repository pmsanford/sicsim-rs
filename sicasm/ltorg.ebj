HLTOR  000000000015
T000000180320066F20063F20064869214142437720033F2FFD444546
E000000
