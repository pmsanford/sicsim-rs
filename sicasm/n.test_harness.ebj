HHARN  000200000006
T000200064B2FFD3F2FFD
E000200
