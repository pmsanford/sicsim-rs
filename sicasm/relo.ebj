HRELO  000000000016
T000000120310000D1B20090F20093F2FFD00000A000005
M00000105
E000000
