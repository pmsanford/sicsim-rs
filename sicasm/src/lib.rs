use anyhow::Result;
use pass_one::FirstPass;
use pass_two::pass_two;
use std::fmt::Write;

mod constants;
mod directive;
mod labels;
mod pass_one;
mod pass_two;
mod record;

/// Assemble a SIC program from text.
///
/// # Errors
///
/// If there's an error in the assembly code
pub fn assemble_program(program_text: &str) -> Result<String> {
    Ok(assemble_with_debug(program_text)?.0)
}

pub fn assemble_with_debug(program_text: &str) -> Result<(String, String)> {
    let lines = program_text.lines().collect::<Vec<_>>();
    let pass_one = FirstPass::parse_lines(&lines)?;

    let (records, debug) = pass_two(pass_one)?;

    let mut assembled = String::new();

    for record in records {
        writeln!(&mut assembled, "{}", record)?;
    }

    Ok((assembled, debug.to_string()?))
}
