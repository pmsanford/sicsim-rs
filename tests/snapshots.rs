use sicasm::assemble_program;

#[test]
fn test_ltorg() {
    let program_text = include_str!("../programs/ltorg.sic");
    let assembled = assemble_program(program_text).unwrap();

    insta::assert_yaml_snapshot!(assembled);
}

#[test]
fn test_copy_xe() {
    let program_text = include_str!("../programs/copy.sic");
    let assembled = assemble_program(program_text).unwrap();

    insta::assert_yaml_snapshot!(assembled);
}

#[test]
fn test_org() {
    let program_text = include_str!("../programs/org.sic");
    let assembled = assemble_program(program_text).unwrap();

    insta::assert_yaml_snapshot!(assembled);
}

#[test]
fn test_word_label_vals() {
    let program_text = include_str!("../programs/word_label.sic");
    let assembled = assemble_program(program_text).unwrap();

    insta::assert_yaml_snapshot!(assembled);
}

#[test]
fn test_equ() {
    let program_text = include_str!("../programs/equ.sic");
    let assembled = assemble_program(program_text).unwrap();

    insta::assert_yaml_snapshot!(assembled);
}

#[test]
fn test_relocate() {
    let program_text = include_str!("../programs/relocatable.sic");
    let assembled = assemble_program(program_text).unwrap();

    insta::assert_yaml_snapshot!(assembled);
}
