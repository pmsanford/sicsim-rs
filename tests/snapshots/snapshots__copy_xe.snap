---
source: tests/snapshots.rs
expression: assembled
---
"HCOPY  000000001076\nT0000001B17202D6900334B1010360320262900003320074B10105D3F2FEC032010\nT00001D120F20160100030F200D4B10105D3E2003454F46\nT00103621B410B400B44075101000E32019332FFADB2013A00433200857C003B850\nT0010531F3B2FEA1340004F0000F1B410774000E32011332FFA53C003DF2009B850\nT001070083B2FEF4F00000505\nM00000705\nM00001405\nM00002705\nM00105A03\nM00107403\nE000000\n"

